File: tests/omp/taskyield1.cpp
#include <omp.h>
#include <stdio.h>

void something_useful ();
void something_critical ();
void foo ( omp_lock_t * lock, int n )
{
   int i;

   for ( i = 0; i < n; i++ ){
      something_useful();
      something_critical();
    }
}

File: tests/omp/taskyield1.ml
open Optitrust
open Prelude


let _ = Run.script_cpp (fun _ ->

  !! Sequence_basic.insert ~reparse:true (stmt "while ( !omp_test_lock(lock) ) {\n
          printf(\"Do something\");}") [tAfter; cCall "something_useful"];
  !! Omp.taskyield [cCall "printf"];
  !! Omp.unset_lock "lock" [tAfter; cCall "something_critical"];
)
